src/test/java/net/snowflake/client/jdbc/ConnectionLatestIT.java

```diff
@@ -720,176 +721,298 @@ public void testKeyPairFileDataSourceSerialization() throws Exception {
+    testPrivateKeyInConnectionString();
+  }
+
+  /**
+   * Works in > 3.18.0
+   *
+   * <p>This test may be split but let's keep it with multiple keys checks to not slow down test
+   * executions
+   */
+  @Test
+  @ConditionalIgnoreRule.ConditionalIgnore(condition = RunningOnGithubAction.class)
+  public void testPrivateKeyBase64InConnectionString() throws SQLException, IOException {
+    Map<String, String> parameters = getConnectionParameters();
+    String testUser = parameters.get("user");
+    String baseUri = parameters.get("uri");
+    try {
+      // Test with non-password-protected private key file (.pem)
+      setUpPublicKey("rsa_key.pub", testUser);
+
+      Properties properties = preparePropertiesForPrivateKeyTests(parameters);
+
+      // PKCS #8
+      String privateKeyBase64 = readPrivateKeyFileToBase64Content("rsa_key.p8");
+      String uri = uriWithPrivateKeyBase64(baseUri, privateKeyBase64);
+      connectSuccessfully(uri, properties);
+
+      // PKCS #1
+      privateKeyBase64 = readPrivateKeyFileToBase64Content("rsa_key.pem");
+      uri = uriWithPrivateKeyBase64(baseUri, privateKeyBase64);
+      connectSuccessfully(uri, properties);
+
+      // test with password-protected private key file (.p8)
+      setUpPublicKey("encrypted_rsa_key.pub", testUser);
+
+      privateKeyBase64 = readPrivateKeyFileToBase64Content("encrypted_rsa_key.p8");
+      uri = uriWithPrivateKeyBase64AndPassword(baseUri, privateKeyBase64, "test");
+      connectSuccessfully(uri, properties);
+
+      // test with incorrect password for private key
+      uri = uriWithPrivateKeyBase64AndPassword(baseUri, privateKeyBase64, "wrong_password");
+      connectExpectingInvalidOrUnsupportedPrivateKey(uri, properties);
+
+      // test with invalid public/private key combo (using 1st public key with 2nd private key)
+      setUpPublicKey("rsa_key.pub", testUser);
+
+      privateKeyBase64 = readPrivateKeyFileToBase64Content("encrypted_rsa_key.p8");
+      uri = uriWithPrivateKeyBase64AndPassword(baseUri, privateKeyBase64, "test");
+      connectExpectingError390144(uri, properties);
+
+      // test with invalid private key
+      privateKeyBase64 = readPrivateKeyFileToBase64Content("invalid_private_key.pem");
+      uri = uriWithPrivateKeyBase64(baseUri, privateKeyBase64);
+
+      connectExpectingInvalidOrUnsupportedPrivateKey(uri, properties);
+    } finally {
+      // clean up
+      unsetPublicKey(testUser);
     }
+  }
+
+  private static String uriWithPrivateKeyBase64(String baseUri, String privateKeyBase64) {
+    return baseUri
+        + "/?"
+        + SFSessionProperty.PRIVATE_KEY_BASE64.getPropertyKey()
+        + "="
+        + privateKeyBase64;
+  }
+
+  private static String uriWithPrivateKeyBase64AndPassword(
+      String baseUri, String privateKeyBase64, String password) {
+    return uriWithPrivateKeyBase64(baseUri, privateKeyBase64)
+        + "&"
+        + SFSessionProperty.PRIVATE_KEY_FILE_PWD.getPropertyKey()
```

Comment:
It is not a file password but a password for base64 content. I suggest adding a new property like `PRIVATE_KEY_BASE64_PWD`. Otherwise, it might be confusing.

Reply:
Marked the FILE_PWD property as deprecated and introduced `PRIVATE_KEY_PWD` for both file and base64 encoded keys